File: backend/services/calibration.py
"""入力テスト（キャリブレーション）の算出（0019）。

frontend が録った生 PCM をそのまま受け取り、**録音時と同じコード**
（services.audio_levels / services.input_profile）で評価する。
UI 側で別実装にすると、テストの判定と実際の挙動がずれる。

数値はすべて **dBFS**（フルスケール基準の相対値）。
物理音圧 dB SPL ではないので、UI でも必ず dBFS と表記すること。
"""
from typing import Optional

import numpy as np

from .audio_levels import (
    ABSOLUTE_SILENCE_RMS,
    CLIP_RATIO_WARN,
    GOOD_SNR_DB,
    HEAVY_GAIN_DB,
    MAX_GAIN_DB,
    MIN_SNR_DB,
    TARGET_SPEECH_DBFS,
    LevelStats,
    amplitude_from_db,
    analyze_levels,
    classify_verdict,
    db_from_amplitude,
    estimate_pass_ratio,
    frame_rms,
    pcm16_to_float32,
    rms,
    recommended_gain_db,
    required_gain_db,
)
from .input_profile import (
    GAIN_AUTO,
    GAIN_MANUAL,
    InputProfile,
    build_profile,
)
from .pcm_stream import SAMPLE_RATE

# 「文字起こしへ確実に回るレベル」の目安。目標発話レベルから 12 dB 下。
# 実測では補正後の発話フレーム中央値が -25.1 dBFS だったので、
# -36 dBFS を下回る発話は補正しても取りこぼしが出やすい。
TRANSCRIBABLE_DBFS = TARGET_SPEECH_DBFS - 12.0
TRANSCRIBABLE_RMS = amplitude_from_db(TRANSCRIBABLE_DBFS)

# 入力テストの想定時間。frontend の案内文と合わせる。
NOISE_STEP_SECONDS = 5.0
SPEECH_STEP_SECONDS = 18.0

TEST_SENTENCE = (
    "これはKoeNoteのマイク入力テストです。\n"
    "会議中と同じ声量で、普段どおりに話しています。\n"
    "離れた位置からでも、音声が正しく認識されるか確認します。"
)


def _current_gain_db(profile: InputProfile, stats: LevelStats) -> float:
    """現在の設定で実際に掛かるゲイン。判定を「現状のまま」で行うために使う。"""
    if profile.gain_mode == GAIN_MANUAL:
        return float(profile.manual_gain_db)
    if profile.gain_mode == GAIN_AUTO:
        return min(recommended_gain_db(stats, max_gain_db=profile.max_gain_db), profile.max_gain_db)
    return 0.0


def analyze_calibration(
    speech_pcm: bytes,
    *,
    noise_pcm: Optional[bytes] = None,
    device_label: str = "",
    profile_payload: Optional[dict] = None,
    sample_rate: int = SAMPLE_RATE,
) -> dict:
    """環境音と遠方発話の 2 ステップから、入力の評価と推奨設定を返す。

    ``noise_pcm`` があればそのノイズフロアを発話測定へ持ち込む。
    無ければ発話測定そのものの 10%ile を使う（片方だけでも動くようにする）。
    """
    profile = build_profile(profile_payload, device_label)

    noise_stats = None
    noise_floor = None
    if noise_pcm:
        noise_samples = pcm16_to_float32(noise_pcm)
        if noise_samples.size:
            noise_stats = analyze_levels(noise_samples, sample_rate=sample_rate)
            noise_floor = _noise_floor_from(noise_samples, sample_rate)

    speech_samples = pcm16_to_float32(speech_pcm)
    stats = analyze_levels(speech_samples, sample_rate=sample_rate, noise_floor=noise_floor)

    gain_db = _current_gain_db(profile, stats)
    recommended_db = recommended_gain_db(stats, max_gain_db=profile.max_gain_db or MAX_GAIN_DB)
    required_db = required_gain_db(stats)

    current_pass = estimate_pass_ratio(stats, gain_db, threshold_rms=TRANSCRIBABLE_RMS)
    recommended_pass = estimate_pass_ratio(stats, recommended_db, threshold_rms=TRANSCRIBABLE_RMS)
    verdict = classify_verdict(
        stats,
        pass_ratio=recommended_pass,
        required_db=required_db,
        max_gain_db=MAX_GAIN_DB,
    )

    # 推奨ゲインを掛けたときにクリップしうるか。ピークで見る。
    projected_peak = stats.peak * amplitude_from_db(recommended_db)
    if projected_peak >= 1.0:
        clipping_risk = "high"
    elif projected_peak >= 0.95:
        clipping_risk = "medium"
    else:
        clipping_risk = "low"

    return {
        "device_label": device_label,
        "input_mode": profile.mode,
        "detected_mode": profile.detected_mode,
        "requested_mode": profile.requested_mode,
        "gain_mode": profile.gain_mode,
        "unit": "dBFS",
        "noise": None if noise_stats is None else noise_stats.as_dict(),
        "speech": stats.as_dict(),
        "noise_floor_dbfs": round(db_from_amplitude(stats.noise_floor), 1),
        "speech_median_dbfs": round(db_from_amplitude(stats.speech_median), 1),
        "speech_low_dbfs": round(db_from_amplitude(stats.speech_low), 1),
        "speech_peak_dbfs": round(db_from_amplitude(stats.peak), 1),
        "snr_db": round(stats.snr_db, 1),
        "speech_ratio": round(stats.speech_ratio, 4),
        "current_gain_db": round(gain_db, 1),
        "current_pass_ratio": round(current_pass, 4),
        "current_skip_ratio": round(1.0 - current_pass, 4),
        "recommended_gain_db": round(recommended_db, 1),
        "recommended_pass_ratio": round(recommended_pass, 4),
        "recommended_skip_ratio": round(1.0 - recommended_pass, 4),
        "required_gain_db": round(required_db, 1),
        "max_gain_db": round(MAX_GAIN_DB, 1),
        "target_speech_dbfs": round(TARGET_SPEECH_DBFS, 1),
        "transcribable_dbfs": round(TRANSCRIBABLE_DBFS, 1),
        "clipping_risk": clipping_risk,
        "clip_ratio": round(stats.clip_ratio, 6),
        "verdict": verdict,
        "thresholds": {
            "good_snr_db": GOOD_SNR_DB,
            "min_snr_db": MIN_SNR_DB,
            "heavy_gain_db": HEAVY_GAIN_DB,
            "clip_ratio_warn": CLIP_RATIO_WARN,
            "absolute_silence_rms": ABSOLUTE_SILENCE_RMS,
        },
        "recommended_settings": recommended_settings(profile, stats, recommended_db),
    }


def _noise_floor_from(samples: np.ndarray, sample_rate: int) -> float:
    """環境音ステップからノイズフロアを求める。

    「話さずに待つ」前提なので、全フレームの**中央値**を採る。
    10%ile だと一瞬の静寂へ引っ張られ、実際の定常騒音より低く出る。
    """
    frames = frame_rms(samples)
    if frames.size == 0:
        return max(rms(samples), 0.0)
    return float(np.median(frames))


def recommended_settings(
    profile: InputProfile, stats: LevelStats, recommended_db: float
) -> dict:
    """「推奨設定を適用」で書き戻す値。

    自動補正で足りるなら auto のまま（実入力に追従する方が固定より強い）。
    最大ゲインだけを必要量へ引き上げる。
    """
    needed = min(max(recommended_db, 0.0), MAX_GAIN_DB)
    # 実運用では話者が近づくこともあるので、必要量そのままではなく少し余裕を持たせる。
    max_gain = min(MAX_GAIN_DB, max(needed + 3.0, 6.0))
    return {
        "inputMode": profile.mode,
        "gainMode": GAIN_AUTO if profile.mode != "loopback" else "none",
        "maxGainDb": round(max_gain, 1),
        "manualGainDb": 0.0,
        "lowInputWarning": True,
    }

File: frontend/src/features/audio/calibration.ts
/**
 * 入力テスト（キャリブレーション）の進行と結果表示（0019）。
 *
 * 判定そのものは Backend（`/api/audio/analyze`）が録音時と同じコードで行う。
 * ここはステップ進行と、結果を人が読める形にする責務だけを持つ。
 *
 * 重要:
 *   - 通常の録音セッションを開始しない。ユーザーの保存先にファイルを残さない
 *   - 数値は必ず **dBFS**（フルスケール基準）と表記する。dB SPL ではない
 */

import type { InputProfileSettings } from './inputProfile';

export type CalibrationStep = 'idle' | 'noise' | 'speech' | 'analyzing' | 'result' | 'error';

/** ステップの想定時間。Backend の calibration.py と揃える。 */
export const NOISE_STEP_SECONDS = 5;
export const SPEECH_STEP_SECONDS = 18;

export const TEST_SENTENCE = [
  'これはKoeNoteのマイク入力テストです。',
  '会議中と同じ声量で、普段どおりに話しています。',
  '離れた位置からでも、音声が正しく認識されるか確認します。'
].join('\n');

export const STEP_TITLES: Record<CalibrationStep, string> = {
  idle: '入力テスト',
  noise: 'ステップ1 / 2　環境音の測定',
  speech: 'ステップ2 / 2　遠方発話の測定',
  analyzing: '解析中',
  result: 'テスト結果',
  error: '入力テストを実行できません'
};

export const STEP_INSTRUCTIONS: Record<CalibrationStep, string> = {
  idle: 'マイクの入力が文字起こしに十分かを測定します。所要 30 秒ほどです。',
  noise: '話さずにお待ちください。部屋の環境音を測定します。',
  speech:
    '実際の会議で最も遠い席へ移動し、普段どおりの声量で下の文を読んでください。',
  analyzing: '測定した音声を解析しています。',
  result: '',
  error: ''
};

/** 総合判定コード。Backend の classify_verdict と一致させる。 */
export type Verdict =
  | 'good'
  | 'usable'
  | 'needs_adjust'
  | 'too_quiet'
  | 'noisy'
  | 'clipping'
  | 'no_input';

export const VERDICT_LABELS: Record<Verdict, string> = {
  good: '良好',
  usable: '使用可能',
  needs_adjust: '要調整',
  too_quiet: '入力が小さすぎる',
  noisy: '環境音が大きすぎる',
  clipping: '音割れの可能性',
  no_input: '入力を検出できない'
};

export const VERDICT_TONES: Record<Verdict, 'ok' | 'warn' | 'error'> = {
  good: 'ok',
  usable: 'ok',
  needs_adjust: 'warn',
  too_quiet: 'error',
  noisy: 'warn',
  clipping: 'warn',
  no_input: 'error'
};

export const VERDICT_ADVICE: Record<Verdict, string> = {
  good: 'このまま録音できます。',
  usable: '自動補正により文字起こしできます。マイクを話者へ近づけると、さらに安定します。',
  needs_adjust: 'マイクの位置を話者へ近づけるか、推奨設定を適用してください。',
  too_quiet:
    '自動補正の上限でも文字起こしに必要なレベルへ届きません。マイクを話者へ近づけるか、外部マイクの使用を検討してください。',
  noisy:
    '話し声に対して環境音が大きすぎます。騒音源を減らすか、マイクを話者へ近づけてください。',
  clipping: '入力が大きすぎます。マイクを話者から離すか、入力音量を下げてください。',
  no_input:
    '入力を検出できません。マイクの接続、システム設定のマイク権限、入力デバイスの選択を確認してください。'
};

/** Backend `/api/audio/analyze` の応答。 */
export interface CalibrationResult {
  device_label: string;
  input_mode: string;
  detected_mode: string;
  unit: string;
  noise_floor_dbfs: number;
  speech_median_dbfs: number;
  speech_low_dbfs: number;
  speech_peak_dbfs: number;
  snr_db: number;
  speech_ratio: number;
  current_gain_db: number;
  current_pass_ratio: number;
  current_skip_ratio: number;
  recommended_gain_db: number;
  recommended_pass_ratio: number;
  recommended_skip_ratio: number;
  required_gain_db: number;
  max_gain_db: number;
  clipping_risk: 'low' | 'medium' | 'high';
  clip_ratio: number;
  verdict: Verdict;
  recommended_settings: Partial<InputProfileSettings> & Record<string, unknown>;
}

/** dBFS 表示。**必ず dBFS と書く**（物理音圧 dB SPL と混同させない）。 */
export function formatDbfs(value: number | null | undefined): string {
  const v = Number(value);
  if (!Number.isFinite(v) || v <= -119) return '−∞ dBFS';
  return `${v.toFixed(1)} dBFS`;
}

/** SNR は 2 点の差なので dB（相対量）で表記する。 */
export function formatDb(value: number | null | undefined): string {
  const v = Number(value);
  if (!Number.isFinite(v)) return '—';
  return `${v >= 0 ? '+' : ''}${v.toFixed(1)} dB`;
}

export function formatPercent(ratio: number | null | undefined): string {
  // Number(null) は 0 になるため、null/undefined は先に弾く。
  if (ratio === null || ratio === undefined) return '—';
  const v = Number(ratio);
  if (!Number.isFinite(v)) return '—';
  return `${(v * 100).toFixed(0)}%`;
}

export const CLIPPING_RISK_LABELS: Record<'low' | 'medium' | 'high', string> = {
  low: '低い',
  medium: '中',
  high: '高い'
};

export interface ResultRow {
  label: string;
  value: string;
  /** 詳細設定側にだけ出す行（一般利用者には数値を見せない）。 */
  advanced?: boolean;
}

/**
 * 結果画面に並べる行。
 *
 * 最大値・最小値だけでは判定していないことが読み取れるよう、
 * 中央値・パーセンタイル・比率を明示する。
 */
export function buildResultRows(result: CalibrationResult): ResultRow[] {
  return [
    { label: '入力デバイス', value: result.device_label || '既定の入力デバイス' },
    { label: '判定された入力モード', value: describeMode(result) },
    { label: 'ノイズフロア', value: formatDbfs(result.noise_floor_dbfs), advanced: true },
    { label: '発話レベル（中央値）', value: formatDbfs(result.speech_median_dbfs) },
    { label: '小さい発話（下位20%）', value: formatDbfs(result.speech_low_dbfs) },
    { label: '発話ピーク', value: formatDbfs(result.speech_peak_dbfs), advanced: true },
    { label: 'SNR', value: formatDb(result.snr_db) },
    { label: '現在設定での推定通過率', value: formatPercent(result.current_pass_ratio) },
    { label: '推定スキップ率', value: formatPercent(result.current_skip_ratio) },
    { label: '必要な推奨ゲイン', value: formatDb(result.recommended_gain_db) },
    {
      label: 'クリッピングリスク',
      value: CLIPPING_RISK_LABELS[result.clipping_risk] ?? '—'
    }
  ];
}

function describeMode(result: CalibrationResult): string {
  const labels: Record<string, string> = { mic: 'マイク', loopback: '内部音声', custom: 'カスタム' };
  const mode = labels[result.input_mode] ?? result.input_mode;
  if (result.input_mode === result.detected_mode) return `${mode}（自動判定）`;
  const detected = labels[result.detected_mode] ?? result.detected_mode;
  return `${mode}（手動選択・自動判定は ${detected}）`;
}

/** 推奨設定を適用したときの改善見込み。再測定前でも見せる。 */
export function describeImprovement(result: CalibrationResult): string {
  const before = formatPercent(result.current_pass_ratio);
  const after = formatPercent(result.recommended_pass_ratio);
  if (result.recommended_pass_ratio <= result.current_pass_ratio + 1e-9) {
    return `現在の設定で既に推定通過率 ${before} です。`;
  }
  return `推定通過率 ${before} → ${after} に改善する見込みです。`;
}

/** 再測定の結果が改善したかを判定する。 */
export function compareRuns(
  before: CalibrationResult,
  after: CalibrationResult
): { improved: boolean; summary: string } {
  const passUp = after.current_pass_ratio > before.current_pass_ratio + 1e-9;
  const clipSafe = riskRank(after.clipping_risk) <= riskRank(before.clipping_risk);
  const improved = passUp && clipSafe;
  const parts = [
    `推定通過率 ${formatPercent(before.current_pass_ratio)} → ${formatPercent(after.current_pass_ratio)}`,
    `クリッピングリスク ${CLIPPING_RISK_LABELS[before.clipping_risk]} → ${CLIPPING_RISK_LABELS[after.clipping_risk]}`
  ];
  return { improved, summary: parts.join(' / ') };
}

function riskRank(risk: 'low' | 'medium' | 'high'): number {
  return { low: 0, medium: 1, high: 2 }[risk] ?? 0;
}

/** 推奨設定を現在の設定へ反映する。未知のキーは無視する。 */
export function applyRecommended(
  current: InputProfileSettings,
  recommended: Record<string, unknown>
): InputProfileSettings {
  const next: InputProfileSettings = { ...current };
  if (typeof recommended.inputMode === 'string') {
    next.inputMode = recommended.inputMode as InputProfileSettings['inputMode'];
  }
  if (typeof recommended.gainMode === 'string') {
    next.gainMode = recommended.gainMode as InputProfileSettings['gainMode'];
  }
  if (Number.isFinite(Number(recommended.maxGainDb))) {
    next.maxGainDb = Number(recommended.maxGainDb);
  }
  if (Number.isFinite(Number(recommended.manualGainDb))) {
    next.manualGainDb = Number(recommended.manualGainDb);
  }
  if (typeof recommended.lowInputWarning === 'boolean') {
    next.lowInputWarning = recommended.lowInputWarning;
  }
  return next;
}

/** ステップの残り秒数から進捗 0..1 を作る。 */
export function stepProgress(step: CalibrationStep, elapsedMs: number): number {
  const total =
    step === 'noise' ? NOISE_STEP_SECONDS : step === 'speech' ? SPEECH_STEP_SECONDS : 0;
  if (total <= 0) return 0;
  return Math.min(1, Math.max(0, elapsedMs / (total * 1000)));
}

/** カウントダウン表示（秒）。 */
export function remainingSeconds(step: CalibrationStep, elapsedMs: number): number {
  const total =
    step === 'noise' ? NOISE_STEP_SECONDS : step === 'speech' ? SPEECH_STEP_SECONDS : 0;
  if (total <= 0) return 0;
  return Math.max(0, Math.ceil(total - elapsedMs / 1000));
}

/** PCM16 の ArrayBuffer 群を base64 へ。Backend へはこの形で送る。 */
export function encodePcmChunks(chunks: ArrayBuffer[]): string {
  let total = 0;
  for (const chunk of chunks) total += chunk.byteLength;
  const merged = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    merged.set(new Uint8Array(chunk), offset);
    offset += chunk.byteLength;
  }
  // 大きい配列を一度に apply するとスタックが溢れるので分割する。
  let binary = '';
  const STEP = 0x8000;
  for (let i = 0; i < merged.length; i += STEP) {
    binary += String.fromCharCode(...merged.subarray(i, i + STEP));
  }
  return btoa(binary);
}
